### PR TITLE
85 read from stdin

### LINKED ISSUE #85: read project from stdin
Adding `-` to mean stdin we can ise flict in pipes, like this

```
tool-that-gathers-info . | flict verify --project-file - 
```

### PR DESCRIPTION
**Pull request by hesa**
Closes #85 

Adding - to mean stdin we can use flict in pipes, like this
```
tool-that-gathers-info . | flict verify --project-file - 
```

### CODE REVIEW

File: flict/flictlib/project.py [hunk 1]
```diff
@@ -68,23 +69,34 @@ def __init__(self, project_file, license_handler, license_expression=None):
...
 
+    # get JSON data from self.project_file
+    def _get_json_data(self):
+        if self.project_file == "-":
+            data = ""
+            for line in sys.stdin:
+                data += line
+            return json.loads(data)
+        else:
+            with open(self.project_file) as fp:
```
Comment (priv-kweihmann) [inline flict/flictlib/project.py]: While you are at it... how about adding some error handling here and there. e.g. `JSONDecodeError` for all cases and `FileNotFound` for the non-stdin way?
Or are you a fan of lengthy hard to read stack traces, eh eh eh? :-)

File: flict/flictlib/project.py [hunk 2]
```diff
@@ -68,23 +69,34 @@ def __init__(self, project_file, license_handler, license_expression=None):
             self.tot_combinations = None
             # self.expand_projects()
 
+    # get JSON data from self.project_file
+    def _get_json_data(self):
+        if self.project_file == "-":
+            data = ""
+            for line in sys.stdin:
```
Comment (priv-kweihmann) [inline flict/flictlib/project.py]: and additionally please change this whole segment here to
```
return json.load(sys.stdin)
```
no need to read line by line, as sys.stdin is already a file pointer

File: flict/flictlib/project.py [hunk 3]
```diff
@@ -68,23 +69,34 @@ def __init__(self, project_file, license_handler, license_expression=None):
...
-                self.project = self.project_object['project']
-            elif 'component' in self.project_object:
-                self.project = self.project_object['component']
-            else:
-                return None
-            if 'meta' in self.project_object:
-                self.meta = self.project_object['meta']
+        self.project_object = self._get_json_data()
+
+        if 'project' in self.project_object:
```
Comment (priv-kweihmann) [inline flict/flictlib/project.py]: Improvement suggestion
```
self.project = self.project_object.get('project', self.project_object.get('component', None))
```
no need to overuse if/else here

File: flict/flictlib/project.py [hunk 4]
```diff
@@ -68,23 +69,34 @@ def __init__(self, project_file, license_handler, license_expression=None):
...
+            self.project = self.project_object['project']
+        elif 'component' in self.project_object:
+            self.project = self.project_object['component']
+        else:
+            return None
+        if 'meta' in self.project_object:
```
Comment (priv-kweihmann) [inline flict/flictlib/project.py]: Same here `self.meta = self.project_object.get('meta', None)` is way more pythonic

File: flict/flictlib/project.py (hunk 1 above)
Comment (hesa) [reply flict/flictlib/project.py]: :+1:

File: flict/flictlib/project.py (hunk 2 above)
Comment (hesa) [reply flict/flictlib/project.py]: :+1:

File: flict/flictlib/project.py (hunk 4 above)
Comment (hesa) [reply flict/flictlib/project.py]: :+1:

File: flict/flictlib/project.py (hunk 3 above)
Comment (hesa) [reply flict/flictlib/project.py]: :+1:

**Comment by priv-kweihmann** [comment]
LGTM - now you just need to rebase and resolve the conflict and we can merge this fella :partying_face:

**Comment by priv-kweihmann** [comment]
Still not properly rebased - GH still showing conflicts :-(

File: flict/flictlib/project.py [hunk 5]
```diff
@@ -68,23 +71,31 @@ def __init__(self, project_file, license_handler, license_expression=None):
...
+        try:
+            if self.project_file == "-":
+                file_name = "sys.stdin"
+                return json.load(sys.stdin)
+            else:
+                file_name = self.project_file
+                with open(self.project_file) as fp:
+                    return json.load(fp)
+        except json.JSONDecodeError:
+            raise FlictException(ReturnCodes.RET_INVALID_PROJECT, "File \"" + file_name + "\" does not contain valid JSON data")
```
Comment (jens-erdmann) [inline flict/flictlib/project.py:85]: `f'File "{file_name}" does not contain valid JSON data'`

File: flict/flictlib/project.py [hunk 6]
```diff
@@ -68,23 +71,31 @@ def __init__(self, project_file, license_handler, license_expression=None):
...
+        except (FileNotFoundError, IsADirectoryError):
+            raise FlictException(ReturnCodes.RET_FILE_NOT_FOUND, "File \"" + file_name + "\" could not be found or is a directory")
```
Comment (jens-erdmann) [inline flict/flictlib/project.py:87]: `f'File "{file_name}" could not be found or is a directory'`

**Comment by jens-erdmann** [comment]
I took the liberty to fix this branch. See https://github.com/vinland-technology/flict/tree/85-read-from-stdin-fixed and feel free to take this as a replacement. We could also just create a new merge request instead.

EDIT: There still seems to be an issue:
```
ERROR
 + Message: Return values "10" and "12" differs
```

**Comment by hesa** [comment]
It seems to work :)

All tests pass. Have also tried some manual testing.

**Pull request merged**